src/librustc_mir/transform/const_prop.rs

```diff
@@ -572,17 +569,18 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         // in MIR. However, if overflow checking is disabled, then there won't be any
         // `Assert` statement and so we have to do additional checking here.
         if !overflow_check {
-            self.use_ecx(source_info, |this| {
+            if let Some(op) = self.use_ecx(|this| {
                 let l = this.ecx.read_immediate(this.ecx.eval_operand(left, None)?)?;
                 let (_, overflow, _ty) = this.ecx.overflowing_binary_op(op, l, r)?;
 
                 if overflow {
-                    let err = err_panic!(Overflow(op)).into();
-                    return Err(err);
+                    Ok(Some(op))
+                } else {
+                    Ok(None)
                 }
-
-                Ok(())
-            })?;
+            })? {
+                self.report_panic_as_lint(source_info, PanicInfo::Overflow(op))?;
```

Comment:
I am quite puzzled why the bitshifts are handled separately above (ignoring `overflow_check`!), and why they use a totally different lint (`EXCEEDING_BITSHIFTS` instead of `CONST_ERR`).

Reply:
Looks like @oli-obk consolidated this lint from a separate pass in 8c8a4335323c3c5fe917aff4ba4acb22804abf3e

Reply:
Looks like @oli-obk consolidated a lint into this code in 8c8a4335323c3c5fe917aff4ba4acb22804abf3e

Reply:
I think we can just remove the bitshift part here, as an overflowing bitshift should already be detected as overflowing by `overflowing_binary_op`. But I didn't dare touching that code as I had no idea of its purpose (and there are no comments explaining that either).

Let's see what @oli-obk says.

Reply:
The difference to `const_err` is that `exceeding_bitshifts` also happens if the lhs is not a constant. So we do need some separate logic.

Reply:
Oh I see... hm yeah, that makes sense. I will add comments along those lines.

However, it could still share the actual error reporting logic, right?

Also, the question is whether this should be under the `if !overflow_check` guard... I guess it shouldn't because the assert terminator will not be evaluated for `x << 44`, but OTOH this means that we will get duplicate errors for `12 << 44`.

Reply:
> However, it could still share the actual error reporting logic, right?

Yes that would work.

> I guess it shouldn't because the assert terminator will not be evaluated for x << 44, but OTOH this means that we will get duplicate errors for 12 << 44.

That has been happening since my first rustc days. Duplicate errors like this are a pain, but we may also be able to get rid of them by just bailing out early and not actually evaluating the statement if any other lint was reported.

Reply:
Also, I think the fact that this PR passes the test suite shows that we don't have a test checking `exceeding_bitshifts` with an unknown (non-constant LHS) in debug mode... as that should be broken with this.

Strangely, I was unable to obtain a duplicate warning here, and I do not understand why.

Reply:
Turns out the lint is already pretty unreliable, see https://github.com/rust-lang/rust/issues/69021.

Is that really worth the mess it is creating?^^

Reply:
to be honest, I have basically started explicitly ignoring this lint after battling with it for a year in the old const evaluator. I'm fine with people improving it, but I don't want to think about it too much ;)

Reply:
Well, it might be better to just remove it then and move on? It would certainly be simpler.^^

[This proposal](https://github.com/rust-lang/rust/pull/68969#discussion_r377107924) could also help, though.

Reply:
Given that this PR does not make things worse than they already are (https://github.com/rust-lang/rust/issues/69021), I am going to close this sub-discussion.